trinity/trainer/tinker/utils.py

```diff
@@ -0,0 +1,258 @@
+            torch.full((length - len(tensor),), pad_value),
+            tensor,
+        ]
+    )
+
+
+def to_tinker_input(
+    experiences: List[Experience], logger: Logger
+) -> Tuple[List[types.Datum], List[types.ModelInput], List[dict]]:
+    assert len(experiences) > 0, "No experiences provided."
+    if experiences[0].experience_type == "dpo":
+        experiences = split_dpo_experience_to_single_turn(experiences)
+
+    batch = []
+    batch_input_tokens = []
+    model_inputs_list = []
+    for exp in experiences:
+        tokens = exp.tokens
+        input_tokens = tokens.long()
+        prompt_length = exp.prompt_length
+        total_length = len(tokens)  # type: ignore
+        response_length = total_length - prompt_length
+        loss_fn_inputs = {
+            "weights": torch.concat(
+                [
+                    torch.zeros(prompt_length - 1, dtype=torch.float32),
+                    exp.action_mask.float(),
+                ]
+            ),
+            "target_tokens": input_tokens.tolist()[1:],
+        }
+        model_inputs = {
+            "total_length": total_length,
+            "action_mask": exp.action_mask,
+        }
+        if exp.reward is not None or exp.token_level_reward is not None:
+            assert exp.logprobs is not None
+            if exp.token_level_reward is not None:
+                if exp.reward is not None:
+                    logger.warning(
+                        "Both exp.rewards and exp.token_level_rewards are provided. "
+                        "Using exp.token_level_rewards."
+                    )
+                token_level_reward = exp.token_level_reward
+            else:
+                token_level_reward = torch.zeros(response_length, dtype=torch.float32)
+                token_level_reward[-1] = exp.reward
+            model_inputs.update(
+                {
+                    "token_level_scores": token_level_reward,
+                    "old_logprob": exp.logprobs,
+                }
+            )
+        for attr in ["advantages", "returns", "teacher_logprobs"]:
+            if getattr(exp, attr, None) is not None:
+                model_inputs[attr] = getattr(exp, attr)
+        # TODO: if tinker support multi-modal input, we can add it here
+        for custom_field in exp.custom_fields:
+            model_inputs[custom_field.destination_field] = torch.tensor(
+                exp.info[custom_field.source_field],
+                dtype=custom_field.data_type,
+            )
+
+        batch.append(
+            types.Datum(
+                model_input=types.ModelInput.from_ints(tokens=input_tokens.tolist()[:-1]),
+                loss_fn_inputs=loss_fn_inputs,
+            )
+        )
+        batch_input_tokens.append(types.ModelInput.from_ints(input_tokens.tolist()))
+        model_inputs_list.append(model_inputs)
+    return batch, batch_input_tokens, model_inputs_list
+
+
+def compute_data_metrics(batch: List[dict[str, torch.Tensor]]) -> dict:
```

Comment:
Check these method with verl version, reduce duplicate code

Reply:
We can merge similar code in the next PR.